[RISCV] Strength reduce mul by 2^M - 3/5/9

preames opened the pull request:
We can expand these as the three instruction sequence: (sub (shl X, C1), (shXadd X, x))

Note: This is a companion to https://github.com/llvm/llvm-project/pull/88983/files which handles the 2^N - 2^M case, but neither is stacked on the other.  Whichever lands second needs rebased to adjust the shadd check.

preames committed:
[RISCV] Strength reduce mul by 2^M - 3/5/9

We can expand these as the three instruction sequence:
(sub (shl X, C1), (shXadd X, x))

Note: This is a companion to https://github.com/llvm/llvm-project/pull/88983/files which handles the 2^N - 2^M case, but neither is stacked on the other.  Whichever lands second needs rebased to adjust the shadd check.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -13447,7 +13447,6 @@ static SDValue expandMul(SDNode *N, SelectionDAG &DAG,
   case 21:
   case 25:
   case 27:
-  case 29:
```
Why is it removed?

preames replied:
It was a mistake to start with.  We don't actually have a TD pattern for 29.

wangpc-pp commented on llvm/test/CodeGen/RISCV/addimm-mulimm.ll:
```diff
@@ -49,19 +53,24 @@ define i64 @add_mul_combine_accept_a3(i64 %x) {
-; RV32IMB-NEXT:    mulhu a3, a0, a2
-; RV32IMB-NEXT:    add a1, a3, a1
-; RV32IMB-NEXT:    mul a2, a0, a2
+; RV32IMB-NEXT:    mulhu a2, a0, a2
```
We may have latency=2 multiply instructions on a high performance RV64 core, so this may be an regression. But this is a RV32 test (multiply won't be so fast), so I just leave this comment for record.

dtcxzyw replied:
FYI, Syntacore SCR1 has single-cycle integer multiplier :(

preames replied:
A couple things here.

First, I agree that we are likely to want a per processor cost based threshold here.  The only question for me is order of work and relative importance, not net result.

I wasn't aware SCR1 has a single cycle latency.  I glanced at the schedule model for that just now, and given the only latencies set appear to be div/rem, I have to ask.  Are you sure this is actually true?

Second, I'll argue that instruction count != latency.  In this particular case, we have two independent instructions followed by a dependent subtract.  On any core with multiple issue and single cycle shl/add, I'd expect that to have a latency of 2.  Obviously, this argument can be taken to a ridiculous extreme, but it isn't obvious to me that a core with single cycle mul only wants single cycle expansions.  The SCR1 might be a weird example here - in the schedule model it looks like there's only one ALU?  Even then, that's not solely a function of latency.  

For comparison, gcc appears to be defaulting to an expansion budget of at least four instructions for rv64gc.

wangpc-pp replied:
I'm not opposed to this optimization.
For high performance core (the latency will be 3-4 normally, 2 is too extreme), the units for simple ALU should be more than multiplier units, so this optimization should be an uplift.
For RV32 core, they are normally single issue with fewer units and the latency of multiplier can be very small because of low frequency (if you agree), I think this optimization will be a decrease.

> First, I agree that we are likely to want a per processor cost based threshold here.
> For comparison, gcc appears to be defaulting to an expansion budget of at least four instructions for rv64gc. (example: https://godbolt.org/z/zzvqfha8f)

Yeah, I agree too. Actually this is what GCC is doing. The cost of mul is 2 for generic OoO, 1 for size, 4 for rocket and others (https://godbolt.org/z/8nvdbo5s1)
https://github.com/gcc-mirror/gcc/blob/58a0b190a256bd2a184554de0fae0031a614ec67/gcc/config/riscv/riscv.cc#L275-L292
We may add similar subtarget feature in LLVM too.

preames commented:
There's a bug here I need to fix, will post a revised version later this morning.

preames committed:
Remove redundant overflow check

At i64, this discards the carry bit, and thus any power of two detection
only considers the bottom 64 in the result.  I believe this is the correct
result.  For i32, the temporary is 64 bits, so an i32 is sign extended in
the temporary, and this reduces to the i64 case.

preames commented:
> There's a bug here I need to fix, will post a revised version later this morning.

Actually, it appears the original code was correct, and I actually had a redundant check, but skeptical review here is appreciated.  Overflow logic makes my head hurt, so a skeptical second look at this is appreciated.

dtcxzyw reviewed (approved):
LGTM.

preames commented:
> LGTM.

JFYI, this needs a significant rebase.  I'm going to wait until https://github.com/llvm/llvm-project/pull/89789 lands, and then revise this to use SHL_ADD directly.

preames committed:
Merge branch 'main' into riscv-mul-pow2-minus-359

preames committed:
Use SHL_ADD
